src/rt-app.c

```diff
@@ -913,174 +921,210 @@ static int __sched_priority(sched_data_t *sched_data)
-			}
-
-			data->lock_pages = 0; /* forced off for SCHED_OTHER */
+			_set_thread_cfs(data, sched_data);
+			_set_thread_uclamp(data, sched_data);
+			data->lock_pages = 0; /* forced off */
 			break;
-
 		case deadline:
-			log_debug("[%d] setting scheduler %s exec %lu, deadline %lu"
-					" period %lu", data->ind,
-					policy_to_string(sched_data->policy), sched_data->period,
-					sched_data->runtime, sched_data->deadline);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=policy policy=dl runtime=%lu deadline=%lu period=%lu",
-				   sched_data->prio, sched_data->runtime,
-				   sched_data->deadline, sched_data->period);
-
-			tid = gettid();
-			sa_params.size = sizeof(struct sched_attr);
-			sa_params.sched_flags = 0;
-			sa_params.sched_policy = SCHED_DEADLINE;
-			sa_params.sched_priority = __sched_priority(sched_data);
-			sa_params.sched_runtime = sched_data->runtime;
-			sa_params.sched_deadline = sched_data->deadline;
-			sa_params.sched_period = sched_data->period;
-
-			ret = sched_setattr(tid, &sa_params, flags);
-			if (ret != 0) {
-				log_critical("[%d] sched_setattr "
-						"returned %d", data->ind, ret);
-				perror("sched_setattr: failed to set deadline attributes");
-				exit(EXIT_FAILURE);
-			}
-		break;
-
+			_set_thread_deadline(data, sched_data);
+			break;
 		default:
 			log_error("Unknown scheduling policy %d",
-					sched_data->policy);
-
-			exit(EXIT_FAILURE);
-	}
-
-	if (sched_data->policy != deadline &&
-	    (sched_data->util_min != -1 || sched_data->util_max != -1)) {
-		sa_params.size = sizeof(struct sched_attr);
-		sa_params.sched_flags = SCHED_FLAG_KEEP_ALL;
-		tid = gettid();
-
-		if (sched_data->util_min != -1) {
-			sa_params.sched_util_min = sched_data->util_min;
-			sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MIN;
-			log_notice("[%d] setting util_min=%d",
-				   data->ind, sched_data->util_min);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=uclamp util_min=%d",
-				   sched_data->util_min);
-		}
-		if (sched_data->util_max != -1) {
-			sa_params.sched_util_max = sched_data->util_max;
-			sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MAX;
-			log_notice("[%d] setting util_max=%d",
-				   data->ind, sched_data->util_max);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=uclamp util_max=%d",
-				   sched_data->util_max);
-		}
-
-		/*
-		 * We can't change clamp values without valid policy and
-		 * priority values.
-		 */
-		sa_params.sched_policy = sched_data->policy;
-		sa_params.sched_priority = __sched_priority(sched_data);
-
-		ret = sched_setattr(tid, &sa_params, flags);
-		if (ret != 0) {
-			log_critical("[%d] sched_setattr returned %d",
-				     data->ind, ret);
-			perror("sched_setattr: failed to set uclamp value(s)");
+				  sched_data->policy);
 			exit(EXIT_FAILURE);
-		}
 	}
 
+	/* Assign a policy, != same, for the current sched_data */
+	sched_data->policy = __sched_policy(data, sched_data);
```

Comment:
If you put this earlier in the function set_thread_param() (ie before switch (__sched_policy)), you can replace all other calls to __sched_policy() by using directly sched_data->policy because sched_data->policy will always be a valid value.
That's the last remaining point to be fix.
The code is easier to read with your changes

Reply:
Indeed, I can remove that portion now.